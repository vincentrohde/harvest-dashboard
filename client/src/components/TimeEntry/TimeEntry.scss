.TimeEntry {
    display: flex;

    @media (max-width:600px)  {
        p {
            margin-bottom: 0;
        }
    }

    &.active {
        background: #ff6f5e;
        color: #fff;

        &.edit {
            background: #fff;
            color: unset;
        }

        .meta-data-container {
            .meta-data {
                .category {
                    color: #0c00ad;
                }
            }
        }

        .time-container {
            .time {
                color: black;
            }

            .actions-header > i {
                margin-left: 1em;
            }
        }

        .title {
            color: black;
        }
    }

    &.edit {
        .edit-form {
            width: 100%;
            display: flex;
            justify-content: space-between;
            flex-wrap: wrap;
        }

        .edit-task, .edit-project {
            flex: 0 0 48%;
        }

        .edit-notes {
            flex: 1;
            margin: 20px 0 0;
            line-height: 2.5;
            border-radius: 4px;
            border: 1px solid #cccccc;
            padding: 0 10px;
            font-size: 1em;
        }
    }

    .entry-container {
        width: 100%;
        display: flex;
        justify-content: space-between;
    }

    .meta-data-container {
        position: relative;
        flex: 1;

        .title {
            position: absolute;
            bottom: 0;
            font-weight: 400;
            font-size: 18px;
            margin: 10px 0 0;

            @media (max-width:600px)  {
                position: relative;
                bottom: unset;
            }
        }

        .meta-data {
            font-size: 14px;

            span {
                &.category {
                    color: dodgerblue;
                }
            }
        }
    }

    .time-container {
        text-align: right;

        .actions-header {
            .delete, .edit {
                display: inline;
            }
        }

        .date {
            margin-right: 10px;
        }

        .time {
            font-size: 22px;
            color: #ff6f5e;
            margin: 10px 0 0;

            i {
                font-size: 0.9em;
            }
        }
    }
}