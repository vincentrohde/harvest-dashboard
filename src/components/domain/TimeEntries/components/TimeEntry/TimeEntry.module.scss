.TimeEntry {
    display: flex;

    @media (max-width:600px)  {
        p {
            margin-bottom: 0;
        }
    }

    &.active {
        background: #ff6f5e;
        color: #fff;

        &.edit {
            background: #fff;
            color: unset;
        }

        .time-container {
            .time {
                color: black;
            }

            .actions-header > i {
                margin-left: 1em;
            }
        }

        .title {
            color: black;
        }
    }

    &.edit {
        .edit-form {
            width: 100%;
            display: flex;
            justify-content: space-between;
            flex-wrap: wrap;
        }

        .edit-task, .edit-project {
            flex: 0 0 48%;
        }

        .edit-notes {
            flex: 1;
            margin: 20px 0 0;
            line-height: 2.5;
            border-radius: 4px;
            border: 1px solid #cccccc;
            padding: 0 10px;
            font-size: 1em;
        }
    }

    .entry-container {
        width: 100%;
        display: flex;
        justify-content: space-between;
    }

    .time-container {
        text-align: right;
    }
}
