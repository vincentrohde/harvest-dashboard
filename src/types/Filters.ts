export interface FiltersInterface {
    dateRange: string[] | 1;
}
