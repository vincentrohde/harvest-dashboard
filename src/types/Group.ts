export type groupsType = GroupInterface[];

export interface GroupInterface {
    id: number;
    name: string;
}