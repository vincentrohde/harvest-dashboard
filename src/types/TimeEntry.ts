import { ProjectInterface } from './Project';
import { TaskInterface } from './Task';

export type timeEntriesType = TimeEntryInterface[];

export interface ShallowTimeEntryInterface {
    hours: number | string;
    id: number;
    notes: string;
    spent_date: string;
    is_running: boolean;
    project: ProjectInterface;
    task: TaskInterface;
}

export interface TimeEntryInterface extends ShallowTimeEntryInterface {
    hours: number;
}

export interface ShallowSubmissionEntryInterface {
    notes: ShallowTimeEntryInterface['notes'];
    spent_date: ShallowTimeEntryInterface['spent_date'];
    hours: ShallowTimeEntryInterface['hours'];
    project_id: number | string;
    task_id: number | string;
}

export interface TimeEntrySubmissionInterface extends ShallowSubmissionEntryInterface {
    hours: TimeEntryInterface['hours'];
    project_id: number;
    task_id: number;
}