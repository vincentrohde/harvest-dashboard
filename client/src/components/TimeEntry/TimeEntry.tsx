// Typescript
import { TimeEntryInterface } from '../../../interfaces/TimeEntry';

// Libs
import React, { useState } from 'react';
import { Icon } from 'semantic-ui-react';
import moment from 'moment';
import EditForm from '../EditForm/EditForm';

// Redux
import { connect } from 'react-redux';
import { deleteTimeEntry } from '../../stores/actions/timeEntries';

// Services
import { backendService } from '../../lib/BackendService/BackendService';
import { timeService } from '../../lib/TimeService/TimeService';

// Stylings
import './TimeEntry.scss';

interface TimeEntryProps extends TimeEntryInterface {
    deleteTimeEntry: Function;
}

type ButtonEvent = React.MouseEvent;

const TimeEntry = ({
       hours,
       id,
       is_running,
       notes,
       task,
       spent_date,
       project,
       deleteTimeEntry
    }: TimeEntryProps) => {

    const [ isActive, setIsActive ] = useState(is_running);
    const [ isEdit, setIsEdit ] = useState(false);
    const { name: taskName, id: task_id } = task;
    const { id: project_id } = project;
    const hoursAndMinutes = timeService.hoursToHoursAndMinutes(hours);
    spent_date = moment(spent_date).format('DD.MM.YYYY');

    const handleClick = (event: ButtonEvent) => {
        const target = event.target as HTMLElement;

        if (target.classList.contains('js-edit')) {
            setIsEdit(!isEdit);
        }

        if (target.classList.contains('js-delete')) {
            backendService.deleteTimeEntry(id)
                .then(() => deleteTimeEntry(id));
        }

        if (target.classList.contains('js-toggle-active')) {
            setIsActive(!isActive);
        }
    };

    return (
        <div className="TimeEntry entry-container tab-container">
            {isEdit
                ? <EditForm
                    isNewEntry={false}
                    entryData={{
                        id,
                        hours: hoursAndMinutes,
                        notes,
                        spent_date,
                        task_id,
                        project_id }}
                    setIsEdit={setIsEdit}/>
                : (<>
                    <div className="meta-data-container">
                        <p className="meta-data pipes">
                            <span className="category">{taskName}</span>
                            <span>{spent_date}</span>
                        </p>
                        <h3 className="title">{notes}</h3>
                    </div>
                    <div className="time-container" onClick={handleClick}>
                        <div className="actions-header">
                            <Icon className="delete js-delete" name="trash alternate"/>
                            <Icon className="edit js-edit" name="pencil"/>
                        </div>
                        <h3 className="time">
                            <Icon className="js-toggle-active"
                                // @ts-ignore
                                name={`clock${isActive ? '' : ' outline'}`}/>
                            {hoursAndMinutes}
                        </h3>
                    </div>
                </>)
            }
        </div>
    )
}

const mapDispatchToProps = { deleteTimeEntry };

export default connect(null, mapDispatchToProps)(TimeEntry);