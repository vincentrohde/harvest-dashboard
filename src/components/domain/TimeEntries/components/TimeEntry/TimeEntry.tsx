// Libs
import React, {useState} from 'react';
import moment from 'moment';
import classNames from 'classnames';

// Services
import apiService from '@/services/ApiService/ApiService';
import timeService from '@/services/TimeService/TimeService';

// Components
import EditForm from '@/components/domain/EditForm/EditForm';
import MetaData from './components/MetaData/MetaData';
import Actions from './components/Actions/Actions';
import Time from './components/Time/Time';

// Types
import {TimeEntryProps} from './TimeEntry.types';

// Stylings
import styles from './TimeEntry.module.scss';

const TimeEntry = ({data, deleteTimeEntry}: TimeEntryProps) => {
    const [isActive, setIsActive] = useState(data.is_running);
    const [isEdit, setIsEdit] = useState(false);
    let {spent_date} = data;
    const {task, project, id, notes, hours} = data;
    const {name: taskName, id: task_id} = task;
    const {id: project_id} = project;
    const hoursAndMinutes = timeService.hoursToHoursAndMinutes(hours);
    spent_date = moment(spent_date).format('DD.MM.YYYY');

    const handleDelete = () => {
        apiService.deleteTimeEntry(id).then(() => deleteTimeEntry(id));
    };

    const toggleEdit = () => {
        setIsEdit(!isEdit);
    };

    const toggleActive = () => {
        setIsActive(!isActive);
    };

    return (
        <div className={classNames(styles.TimeEntry, styles['entry-container'], 'tab-container')}>
            {isEdit ? (
                <EditForm
                    onCancel={() => setIsEdit(false)}
                    onSuccess={() => setIsEdit(false)}
                    data={{
                        id,
                        hours: hoursAndMinutes,
                        notes,
                        spent_date,
                        task_id,
                        project_id,
                    }}
                    setIsEdit={setIsEdit}
                />
            ) : (
                <>
                    <MetaData notes={notes} date={spent_date} task={taskName} />

                    <div className={styles['time-container']}>
                        <Actions handleDelete={handleDelete} toggleEdit={toggleEdit} />
                        <Time
                            isActive={isActive}
                            hoursAndMinutes={hoursAndMinutes}
                            toggleActive={toggleActive}
                        />
                    </div>
                </>
            )}
        </div>
    );
};

export default TimeEntry;
