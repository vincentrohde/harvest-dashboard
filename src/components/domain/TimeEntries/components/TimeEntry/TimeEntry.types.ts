import { TimeEntryInterface } from '@/types/TimeEntry';

export interface TimeEntryProps {
    data: TimeEntryInterface;
    deleteTimeEntry: Function;
}
