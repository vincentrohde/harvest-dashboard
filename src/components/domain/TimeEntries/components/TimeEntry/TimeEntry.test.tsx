// Libs
import { render } from '@testing-library/react';
import React from 'react';
import '@testing-library/jest-dom/extend-expect';

// Components
import TimeEntry from './TimeEntry';

// Types
import { TimeEntryProps } from './TimeEntry.types';

const mockData: TimeEntryProps = {
    data: {
        hours: 1,
        id: 1,
        notes: 'test note',
        spent_date: '01/01/2022',
        is_running: false,
        project: {
            id: 1,
            name: 'test project'
        },
        task: {
            id: 1,
            name: 'test task'
        }
    },
    deleteTimeEntry: () => null
}

describe('<TimeEntry />', () => {
    it('renders', () => {
        const { asFragment } = render(
            <TimeEntry {...mockData} />
        );

        expect(asFragment()).toMatchSnapshot();
    });
})
