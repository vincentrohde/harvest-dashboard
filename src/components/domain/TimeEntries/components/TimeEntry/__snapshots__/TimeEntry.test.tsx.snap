// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimeEntry /> renders 1`] = `
<DocumentFragment>
  <div
    class="TimeEntry entry-container tab-container"
  >
    <div
      class="meta-data-container"
    >
      <p
        class="meta-data pipes"
      >
        <span
          class="task"
        >
          test task
        </span>
        <span>
          01.01.2022
        </span>
      </p>
      <h3
        class="title"
      >
        test notes
      </h3>
    </div>
    <div
      class="time-container"
    >
      <div
        class="actions-header"
      >
        <i
          aria-hidden="true"
          class="trash alternate icon delete js-delete"
        />
        <i
          aria-hidden="true"
          class="pencil icon edit js-edit"
        />
      </div>
      <h3
        class="Time"
      >
        <i
          aria-hidden="true"
          class="clock outline icon js-toggle-active"
        />
        1:00
      </h3>
    </div>
  </div>
</DocumentFragment>
`;
